Return detached threads to the pool

VirtualTim opened the pull request:
Once detached threads are finished their execution they emit the 'exit' command. Instead of a noop they should rejoin the pool. 
I don't know if this change has any other consequences.

Resolves #8201.

VirtualTim commented:
Well tests are failing because joinable threads are calling exit, which cleans them up. Then the join fails.

As far as I can tell there is no standard way to tell if a thread is joinable from within that thread, so I think that rules out posting different messages (unless we set a flag ourselves).
The alternative would be to return success on joining a thread that doesn't exist (because it's been disposed of). This doesn't seem to go against [the free pthreads spec](http://pubs.opengroup.org/onlinepubs/9699919799/) I could find, but is different behaviour from [Linux pthread_join](http://man7.org/linux/man-pages/man3/pthread_join.3.html), which returns ESRCH in this situation.
I don't know how this would affect std::thread's.

VirtualTim commented:
So seems like my approach does the trick.
@juj does this seem reasonable to you? You seem like the expert on this sort of stuff.
@benjymous if you manually apply my patch to library_pthread.js does this fix your issue?

VirtualTim commented:
@juj (or anyone else) Are you able to review this?

kripken commented:
It looks like we mark a thread as detached if we join it or if it calls detach, is that right?

sbc100 commented on src/library_pthread.js:
```diff
@@ -319,18 +319,18 @@ var LibraryPThread = {
-              // TODO: Free if detached.
-              PThread.runningWorkers.splice(PThread.runningWorkers.indexOf(worker.pthread), 1); // Not a running Worker anymore.
+            } else if (d.cmd === 'exit' || d.cmd === 'cancelDone') {
+              var detached = Atomics.load(HEAPU32, (d.threadId + {{{ C_STRUCTS.pthread.detached }}} ) >> 2);
```
Don't you want to maintain the old behaviour when `cancelDone` is the command?

What about making this a function and calling it unconditionally in `cancelDone` and conditionally during `exit`.

(I'm not an expert on this part of the code so I could well be wrong here, but we seems that you are change the behaviour on cancelDone.. also cancelDone doesn't have threadId property).

VirtualTim replied:
Yeah making that a function might be a smarter idea.

VirtualTim commented:
> It looks like we mark a thread as detached if we join it or if it calls detach, is that right?

It was a while since I did this initially, but if I remember right joined threads are marked as detached after joining, but if you try and join a detached thread it will return an error.

I've since refactored things as per @sbc100's suggestion, and I think it's now a bit clearer.

VirtualTim commented:
For reference this is how I tested this:
```
#include <thread>
#include <iostream>
#include <math.h>
#include <emscripten/html5.h>

void some_function(const EmscriptenMouseEvent* e) {
	std::cout << "Thread ID: " << std::this_thread::get_id() << std::endl;
	double d;
	for (int i=0; i<100000000; i++)			//simulate work
		d += (i%2 ? sqrt((int)(100*rand())) : (-1)*sqrt((int)(100*rand())));
	std::cout << "    Finished work:" << d << std::endl;
}

EMSCRIPTEN_RESULT click_callback(int eventType, const EmscriptenMouseEvent* e, void* userdata) {
	std::thread(some_function, e).detach();
	return EMSCRIPTEN_RESULT_SUCCESS;
}

int main(int argc, char** argv) {
	const char* canvasID = "#canvas";
	emscripten_set_mousedown_callback(canvasID, 0, EM_TRUE, click_callback);
}
```
Compile `like emcc test.cpp -std=c++11 -s USE_PTHREADS=1 -s NO_EXIT_RUNTIME=1 -s TOTAL_MEMORY=838860800 -o test.html`

Clicking the canvas will spawn a detached thread. Pre-PR once the thread was finished it would still exist, and a new thread would get created each time. Post-PR threads will get re-used if they are available, and if no threads are free a new one will be spawned. You can observe this by watching the number of threads listed in the browsers dev tools.

kripken commented:
Thanks, yeah, I think it's clearer now. The code lgtm. Please add a test, though (maybe that code right there, with small modifications)? Can be alongside the other `browser.test_pthread_*` tests - see those for examples (in tests/test_browser.py). Instead of watching the browser devtools for the count, can iterate say 1,000 times - almost certainly browsers will not allow that many actual threads, so we must be reusing them.

kripken commented:
Btw, the test failures look unrelated - merging in latest incoming should fix them, might be an old problem.

VirtualTim commented:
I've got a test working (still need to figure out how to integrate it though), but I did discover another issue. If you create a bunch of threads take a look at `PThread.pthreads`. You'll see this object keeps growing, and contains a reference to every thread created. It doesn't look like those are ever deleted, so I've added that in.

kripken commented:
@VirtualTim sorry for the delay here. Code looks good to me. The test, however looks like it might be racy - the dependence on timing is risky, we may end up with random failures because of it. I think it would be better to rewrite it in a way that does not depend on time measurements (instead, can communicate with mutexes etc. to tell the threads what to do at each time etc.).

VirtualTim commented:
Yeah you're right, it does have the potential to be racy. However I did test this locally with 0.1s intervals instead of 0.5s, and it worked fine. I figured that 5x the wait would make failure due to a race condition very unlikely. Plus it runs with 20 threads, so the chance that even one wouldn't finish seems very unlikely. I guess my motivation for doing it this was way trying to emulate a real-world scenario as closely possible.

So just to check if I'm understanding you correctly, are you suggesting putting a mutex around the lambda inside `spawn_a_thread`, to basically force the threads to run sequentially?

kripken commented:
Well, the mutex comment was just a way to try to make it deterministic - there might be better ways. Like perhaps the threads can communicate using a shared atomic. But the key thing is to not depend on timing for their decisions.

Sorry about this, but we've had many tests that depend on timing and that seem stable, but eventually become random errors because of a change on the browser side. So it's really important to avoid as much nondeterminism as possible.

juj commented on tests/pthread/test_pthread_detach.cpp:
```diff
@@ -0,0 +1,57 @@
+	std::thread( [] {
+		double d=0;
+		for (int i=0; i<10000000; i++)			//simulate work
+			d += (i%2 ? sqrt((int)(rand())) : (-1)*sqrt((int)(rand())));
```
For the purposes of this test, each thread can return immediately without burning CPU overhead - this is because there is a element of timing in play in this test, i.e. if this happens to take a long time or stall in the browser due to some unspecified reason, then potentially all threads might live together.

VirtualTim replied:
Yeah, I guess I was just concerned that spawning a thread that does nothing could be removed by the compiler as an optimisation. And I guess I thought there might be some sort of special handling, or timing issue when trying to observe a thread that terminates instantly.
But you're right, this is probably not the best approach. I'll take another crack at writing a decent test.

juj commented on tests/pthread/test_pthread_detach.cpp:
```diff
@@ -0,0 +1,57 @@
+			setTimeout(() => { _spawn_a_thread(); }, i*500);
+		}
+		
+		setTimeout(() => { _count_threads(i_max); }, i_max*500);
```
This is a clever way to test, though can we make this test much shorter? A 10 second long test is quite long in comparison to most tests in the suite, so these kind of things add up the overall suite runtime. Perhaps this can be tested e.g. by spawning one thread detached, then exiting it, and then when that is confirmed to have exited (e.g. via an atomic var), wait for 500 msecs for good measure, and then run a second thread, which should be run in the context of the same worker that ran the first thread. That way the test should finish in less than a second, and if the test needs to be duplicated in multiple modes in the future (PROXY_TO_PTHREAD & asm.js vs wasm & different opts modes are common different cases that often are needed to test), that will not multiply the overall test suite run by that much.

VirtualTim replied:
Redid this.
Yeah you were right, a 10s test is far too long. It should now take about 1/2s. I didn't use an atomic, because I figured there was going to be some time between a thread going out of scope and being returned to the pool.
It's not completely race free, but I'm performing 5 100ms checks on the thread status, so if even one thread hasn't returned something has probably gone wrong.

juj commented on src/library_pthread.js:
```diff
@@ -239,7 +239,15 @@ var LibraryPThread = {
       if (pthread.worker) pthread.worker.pthread = null;
     },
-
+    returnThreadToPool: function(workerToFree) {
```
Let's rename this to `returnWorkerToPool` (threads are never returned to the pool, threads are quit, and then the worker that was hosting the thread will return to the pool) - also `workerToFree` -> `worker`

VirtualTim replied:
Yeah good point.

juj commented on src/library_pthread.js:
```diff
@@ -320,17 +328,16 @@ var LibraryPThread = {
               alert('Thread ' + d.threadId + ': ' + d.text);
             } else if (d.cmd === 'exit') {
-              // Thread is exiting, no-op here
+              var detached = Atomics.load(HEAPU32, (d.threadId + {{{ C_STRUCTS.pthread.detached }}} ) >> 2);
```
Instead of passing a `d.threadId`, the thread ID of the thread that the given current can be located with `worker.pthread.thread` - that is nicer since it retains more coherency, and less state to postMessage over. I now see that the `alert` command does also pass a `threadId` message (that is actually the only command that does past the `threadId` as custom, but it should not need to, for the same reason)

VirtualTim replied:
Ah good point. I did this before writing the `returnThreadToPool` function, so I didn't realise that at the time. I just followed how the above messages behave. `print`, `printErr` and `alert` all take a `d.threadId`. I'm not going to change that in this PR, but might open another one to address this.

juj commented on src/library_pthread.js:
```diff
@@ -185,7 +185,7 @@ var LibraryPThread = {
 #endif
 
-          postMessage({ cmd: 'exit' });
+          postMessage({ cmd: 'exit', threadId: tb });
```
Passing `threadId: tb` here is not needed, in the `onmessage` the receiver will know which thread the worker was hosting, via `worker.pthread.thread` field.

juj commented on src/library_pthread.js:
```diff
@@ -320,17 +328,16 @@ var LibraryPThread = {
-              // Thread is exiting, no-op here
+              var detached = Atomics.load(HEAPU32, (d.threadId + {{{ C_STRUCTS.pthread.detached }}} ) >> 2);
+              if (detached) {
+                PThread.returnThreadToPool(worker);
```
Perfect! Indeed in this case, if the thread had been detached, the code was not returning the worker to the pool, but only the two cases of `pthread_cancel()` and non-detached threads with `pthread_join()` were returning workers to the pool. Looks good.

juj commented:
Thanks for the PR! Indeed workers hosting detached threads were not properly returning to the worker pool. The fix looks good, only minor comments.

VirtualTim commented:
Thanks for the review juj. I've added the changes you suggested, and rewrote the test. Now it should only take about half a second to run.
The test should be pretty race-free, but I did add a timeout in case a thread takes too long to return to the pool. But better for the test to timeout and fail than hang on failures, right?

Does that all look good now?

VirtualTim commented:
So the failing tests seem to be when `pthread_join` is used on the main thread.
My understanding is that this is pretty much never going to work as `join` will block the main thread, and the worker needs to send a message to the main thread to finish exiting. So this will deadlock more often than not.

The docs even mention that it's not supported: https://emscripten.org/docs/porting/pthreads.html, even though `join` should work when called form a non-main thread.

@juj, is my understanding correct, or should `join` on the main thread actually work?

VirtualTim commented:
Oh whoops, missed the `-s PROXY_TO_PTHREAD=1` flag. No wonder I was having issues getting the test to run. I'll look into it today.

VirtualTim commented on tests/pthread/test_pthread_detach.cpp:
```diff
@@ -0,0 +1,76 @@
+		double d=0;
+		for (int i=0; i<10; i++)			//simulate work
+			d += (i%2 ? sqrt((int)(rand())) : (-1)*sqrt((int)(rand())));
+	} ).detach();
```
Ah right. Yeah I noticed that the other tests were written using pthreads, but I find the sdt thread library easier to use, and I'm more familiar with that, so that's what I went with. So yeah, I can rename the test.

VirtualTim commented:
OK, so looks like that last change broke a bunch of tests on Firefox. I think it might be better to revert the change on this line: https://github.com/emscripten-core/emscripten/pull/8286/files#diff-db41bea94577c2dd9b0eef0308b06cf9R243
and open another PR to fix the leak this was trying to address.

VirtualTim commented:
OK, investigating the failing test reviled a weird issue. The `threadExit` function executes on the worker, and sets the exit status. However for the thread to _actually_ exit it has to post the exit command to the main thread. So there can be unlucky circumstances where the thread marks itself as exited but it hasn't actually exited. I think I know how to fix that.

The other note is that the `test_pthread_mutex.cpp` test joins on the main thread, so will by pretty race-y. I'm surprised I haven't seen it fail more. Unless I'm missing something I think it should be built with `-s PROXY_TO_PTHREAD=1`. But I think that change should be made in a separate PR.

VirtualTim commented:
Ugh, I tried to run some tests with CircleCI, but turns out that doing so hijacked the tests on this page. is someone (@kripken?) re-run the tests on your CI system for me?
Thanks.

kripken commented:
Oddly I can't rerun tests here - it says I don't have write permissions. It's using the permissions for your fork, I guess, and not the upstream repo?

As a workaround, can merge incoming into here.

VirtualTim commented:
OK, going through the test failures:

| Test | Result |
| ------------- | ------------- |
| test_glgears_proxy_jstarget | Firefox specific test. I didn't test this one |
| test_pthread_64bit_atomics | `join`'s on main thread. Should be compiled with `-s PROXY_TO_PTHREAD=1` |
| test_pthread_64bit_cxx11_atomics | Tried compiling with all the options used for the test in both Firefox and Chrome and I couldn't get this to fail |
| test_pthread_atomics | `join`'s on main thread. Should be compiled with `-s PROXY_TO_PTHREAD=1` I didn't test with `modularize`, but I assume this is the same. |
| test_pthread_barrier | `join`'s on main thread. |
| test_pthread_call_async_on_main_thread | Tried compiling with all the options used for the test in both Firefox and Chrome and I couldn't get this to fail |
| test_pthread_call_sync_on_main_thread | Bad tests. I've opened #8621 to fix them. |
| test_pthread_cancel | Worked for me on both Firefox/Chrome |
| test_pthread_cleanup | `join`'s on main thread. Works with `-s PROXY_TO_PTHREAD=1` |
| test_pthread_clock_drift | Worked for me on both Firefox/Chrome |
| test_pthread_condition_variable | `join`'s on main thread. |

So it's a bit odd that some tests (4) work for me, but hang on the build machine. I'll double check my commits, perhaps I missed something.
Also a number of tests call `pthread_join` on the main thread, which was always a bad idea, since that was very flaky. The changes I made remove that flakeyness by setting the exited flag to when the thread is "more exited", but it basically means that joining a thread on the main browser thread will always hang, instead of often hanging. Even though this could break some behaviour, it's more consistent and that behaviour would have been very fragile anyway.
I could go into some more technical details, but a proper explanation would be a few paragraphs.

So what should we do with these tests? The easy fix would be to add `-s PROXY_TO_PTHREAD=1` to the 5 that don't have them, and then remove that setting when we get native WASM threads.
Thoughts?

sbc100 commented:
I don't understand why this change needs those tests to have `PROXY_TO_PTHREAD` where before they didn't?

My understanding is that its possible to do thread creation and joining on the main thread if you have `PROXY_TO_PTHREAD` enabled *or* if you pre-create your threads using `PTHREAD_POOL_SIZE`.  Does this change prevent the latter from working now?  (I'm not saying that would necessarily block this change, just want to get it clear what is changing).

kripken commented on tests/pthread/test_std_thread_detach.cpp:
```diff
@@ -0,0 +1,76 @@
+		
+		//Check if a worker is free every threads_to_spawn*100 ms, or until max_thread_check is exceeded
+		const SpawnMoreThreads = setInterval(() => {
+			if (PThread.unusedWorkerPool.length > 0) {	//Spawn a thread if a worker is available
```
it would be better to write this using just the pthreads API, and not look at `PThread.unusedWorkerPool` and other internal details, since this might change in future refactorings. but if it's much easier to write it this way then it's fine for now I think

VirtualTim replied:
Yeah I'm not sure this can be done from the pthreads API. I mean, you can't really use the API to test the API, right?

I realise that this will likely change once we get native WASM threads, but I think by then we'll need to redo a lot of the pthreads tests anyway.

kripken commented:
Is the discussion of tests here about general improvements, or necessary changes for this PR? The tests all look green now here, so I think this PR can land?

VirtualTim commented:
> My understanding is that its possible to do thread creation and joining on the main thread if you have `PROXY_TO_PTHREAD` enabled or if you pre-create your threads using `PTHREAD_POOL_SIZE`.

`PROXY_TO_PTHREAD` should always work, since the `join` actually happens on a worker, not the main thread. My guess is that `PTHREAD_POOL_SIZE` helps hide the racyness since it's much quicker to pull a thread off the pool than create a new one.

So I wrote up some big explanation about everything, but then realised that it's not actually an issue, since on `join`-ing the 'exit' command does nothing.

Anyway, for a quick overview (using `PROXY_TO_PTHREAD=0`, `join` on main thread):
* Main thread waits in a loop checking thread status.
* Thread exits, setting thread status: https://github.com/emscripten-core/emscripten/pull/8286/files#diff-db41bea94577c2dd9b0eef0308b06cf9R148
  * Then send the 'exit' message to the main thread: https://github.com/emscripten-core/emscripten/pull/8286/files#diff-db41bea94577c2dd9b0eef0308b06cf9R162
However if the main thread is still processing a message (as it is when waiting on a `join`) then the 'exit' message is processed after the thread has finished joining. So you can end up with two issues:
1. The thread has joined, but 'exit' hasn't been executed.
1.1 Currently 'exit' is a noop on joining threads, so this is "OK". Conceptually I don't really like it, but I don't think I can fix it without breaking more stuff.
2. If the thread needs to execute something on the main thread it can't until the thread has finished joining.
2.1 This is why joining on the main thread is super risky.
2.2 This "works" if the main thread receives the message before the join command, but it's really race-y.

I guess I'd like to change this, but I don't really think there is a good way to to it using web workers. I think we're better off waiting for native WASM threads.

VirtualTim commented:
@kripken Yeah I'm happy for this to land. Turns out that this was more complicated than I first thought (which was probably why it wasn't implemented in the first place), but I think everything's as good as it's going to get. 
There's still some weirdness using workers to emulate threads, but I don't think this is really something that can be fixed, and one day we should get native WASM threads anyway.

kripken commented:
Great, thanks, merging.
